### PR TITLE
MTDSA-3084 - Fix new microservices to properly handle bad json body

### PR DESCRIPTION
**Pull request by jeremystone**

### CODE REVIEW

File: app/config/ErrorHandler.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      case Some(VERSION_2) =>
+        Logger.error(s"! onClient error in version 2, for (${request.method}) [${request.uri}] with status: $statusCode and message: $message")
+        statusCode match {
+          case play.mvc.Http.Status.BAD_REQUEST =>
+            auditConnector.sendEvent(dataEvent("ServerValidationError", "Request bad format exception", request))
+            Future.successful(BadRequest(ErrorCode.invalidRequest.asJson))
+
+          case play.mvc.Http.Status.NOT_FOUND =>
+            auditConnector.sendEvent(dataEvent("ResourceNotFound", "Resource Endpoint Not Found", request))
+            Future.successful(
```
Comment (ma3574) [inline app/config/ErrorHandler.scala]: Reformat code to be on one line

File: app/config/ErrorHandler.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                            ) extends JsonErrorHandler(config, auditConnector) {
+
+  private val versionRegex = """application\/vnd.hmrc.(\d.\d)\+json""".r
+
+  override def onClientError(request: RequestHeader, statusCode: Int, message: String): Future[Result] = {
+
+    implicit val headerCarrier: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+    getAPIVersionFromRequest match {
+      case Some(VERSION_2) =>
+        Logger.error(s"! onClient error in version 2, for (${request.method}) [${request.uri}] with status: $statusCode and message: $message")
```
Comment (ma3574) [inline app/config/ErrorHandler.scala]: Should be `Logger.warn`

File: app/config/ErrorHandler.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        statusCode match {
+          case play.mvc.Http.Status.BAD_REQUEST =>
```
Comment (ma3574) [inline app/config/ErrorHandler.scala]: Can we not import ` play.mvc._` and just have `BAD_REQUEST` and `NOT_FOUND`

File: app/config/ErrorHandler.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+            Future.successful(Status(statusCode)(Json.toJson(statusCode match {
+              case UNAUTHORIZED => unauthorisedError
+              case UNSUPPORTED_MEDIA_TYPE => invalidBodyType
+              case _ => ErrorCode("INVALID_REQUEST", message)
+            })))
+        }
+
+      case _ =>
+        Logger.error(s"! onClient error, for (${request.method}) [${request.uri}] with status: $statusCode and message: $message")
```
Comment (ma3574) [inline app/config/ErrorHandler.scala]: `Logger.warn`

Comment (ma3574) [inline app/config/ErrorHandler.scala]: Why do we use an exclamation mark?

File: app/config/ErrorHandler.scala [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+          case _ => (INTERNAL_SERVER_ERROR, internalServerError, "ServerInternalError")
+        }
+
+        auditConnector.sendEvent(
+          dataEvent(eventType, "Unexpected error", request, Map("transactionFailureReason" -> ex.getMessage)))
+
+        Future.successful(Status(status)(Json.toJson(errorCode)))
+
+      case _ =>
+        ErrorHandler.super.onServerError(request, ex)
```
Comment (ma3574) [inline app/config/ErrorHandler.scala]: No new line needed

File: test/config/ErrorHandlerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NoStackTrace
+
+/**
+  * Unit tests for [[ErrorHandler]]
+  */
+class ErrorHandlerSpec extends UnitSpec with Mock with GuiceOneAppPerSuite {
+  implicit val erFormats: OFormat[ErrorResponse] = Json.format[ErrorResponse]
+
+  class DummyRequestHeader(hdrs: (String, String)*) extends RequestHeader {
```
Comment (ma3574) [inline test/config/ErrorHandlerSpec.scala]: Is this DummyRequestHeader needed? Should this not be something that is mocked?

Comment (MarkAKelly) [reply test/config/ErrorHandlerSpec.scala]: You can use `FakeRequest().withHeaders()` for this

File: app/config/ErrorHandler.scala (hunk 2 above)
Comment (MarkAKelly) [reply app/config/ErrorHandler.scala]: Can we use the logging format we've used elsewhere in the microservices?  i.e `[Class] [function]...`

File: app/config/ErrorHandler.scala [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+        val (status, errorCode, eventType) = ex match {
+          case _: NotFoundException => (NOT_FOUND, matchingResourceNotFound, "ResourceNotFound")
+          case _: AuthorisationException => (UNAUTHORIZED, unauthorisedError, "ClientError")
+          case _: JsValidationException => (BAD_REQUEST, invalidRequest, "ServerValidationError")
+          case _ => (INTERNAL_SERVER_ERROR, internalServerError, "ServerInternalError")
+        }
+
+        auditConnector.sendEvent(
+          dataEvent(eventType, "Unexpected error", request, Map("transactionFailureReason" -> ex.getMessage)))
```
Comment (MarkAKelly) [inline app/config/ErrorHandler.scala]: Indent this the same as the previous audit event, i.e

```
auditConnector.sendEvent(
              dataEvent(
                eventType = "ClientError",
                transactionName = s"A client error occurred, status: $statusCode",
                request = request,
                detail = Map.empty
              )
            )
```

Comment (narendravijayarao) [reply app/config/ErrorHandler.scala]: Used the same format and eventType as JsonErrorHandler. Again if we change do we need to inform to the TXM team?

Comment (jeremystone) [reply app/config/ErrorHandler.scala]: Probably need to make it obvious that the handling and audit events in onServerError and onClientError are the same given equivalent scenarios (e.g. NotFoundException ~ status code NOT_FOUND etc...). This is not the case at the moment.

Comment (MarkAKelly) [reply app/config/ErrorHandler.scala]: @narendravijayarao I'm not saying change the event type.  Just indent it in the same way as the other one.

**Pull request merged**